Feature/add email parameter

brendanarnold opened the pull request:
Hi, this is a 'best guess' PR for the mapper to supply email as a parameter which would be really handy for my use case.

I don't have a Java build environment so admit I'm flying blind here but, looking at the docs for the Keycloak classes, it seems that `.getEmail()` would work and its a relatively small change.

I admit it hasn't been tested, or even compiled - please let me know if you would like me to work this up more

brendanarnold committed:
Added email to ExternalClaimMapper.java

Added email as an option to pass - an untested commit

brendanarnold committed:
Update README.md

zloom commented:
Hi Brendan,

Thanks for contributing! Overall, the changes look solid.

One thing needs to be addressed before merging: the email parameter is currently inserted into the URL without encoding. This may break requests when special characters like @, +, or = are present.

Please use `URLEncoder.encode(email, StandardCharsets.UTF_8)` before substitution.

Once that’s fixed, I’ll be happy to approve and merge.

Cheers,
Anton

brendanarnold commented:
Ah that’s a great shout - I’m away from my computer for the next two weeks
but will
update when I get back

Brendan

On Tue, 27 May 2025 at 12:17, Anton Zalialdinov ***@***.***>
wrote:

> *zloom* left a comment (zloom/keycloak-external-claim-mapper#1)
> <https://github.com/zloom/keycloak-external-claim-mapper/pull/1#issuecomment-2911978266>
>
> Hi Brendan,
>
> Thanks for contributing! Overall, the changes look solid.
>
> One thing needs to be addressed before merging: the email parameter is
> currently inserted into the URL without encoding. This may break requests
> when special characters like @, +, or = are present.
>
> Please use URLEncoder.encode(email, StandardCharsets.UTF_8) before
> substitution.
>
> Once that’s fixed, I’ll be happy to approve and merge.
>
> Cheers,
> Anton
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zloom/keycloak-external-claim-mapper/pull/1#issuecomment-2911978266>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEOXUQM2HRI4JF4ZNWMO4D3AQ3UPAVCNFSM6AAAAAB53DO2X6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMJRHE3TQMRWGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Commit:
Encodes email before mapping into URL

brendanarnold commented:
Hi @zloom - I have updated with the `URLEncoder` as requested. 

I have been able to compile it using the Docker build script and the extension is recognised when running the Docker Compose dev setup. However I haven't been able to test the functionality since the dev Realm import failed when I tried to upload it into Keycloak.

The Java code may not be idiomatic as well - its been about 10 years since I did any Java dev!

zloom commented:
You probably need to reconfigure the realm, export it, and update the realm file in the repo. Then the import should work out of the box with the new parameter.

I recommend starting from a fresh environment checked out from master, then making the code changes and updating the extension. That way, you’ll have a realm based on master with the updated extension, which you can export and commit.

Let me know if you get stuck — I’ll be happy to help.

brendanarnold commented:
Hmm, I've been playing around with it - it looks like there is a bug that causes the mapping to fail if the request headers are not set - we get a `java.lang.IllegalArgumentException: argument "content" is null` exception at ...

```var headers = mapperModel.getConfigMap(REQUEST_HEADERS_PROPERTY);```

Including an arbitrary header in the settings fixes it but wondering if this is a known issue? It doesn't have anything to do with my changes as far as I can tell

Commit:
Fix bug when no headers set

Commit:
Add explicit import repository so will work with Podman

Commit:
Update README with test instructions

Commit:
Include instructions for inspecting the JWT

Commit:
Update realm export

Commit:
Correct to URL encoded output in README

brendanarnold commented:
Hi @zloom I've run the plugin locally using the compose script and it worked - there are a couple of other changes in the PR

 * Fix bug where not specifying the headers property causes crash of the plugin
 * Added explicit repository so that the compose script will work with Podman (this is cross-compatible with Docker)
 * Updated README with explicit instructions on how to test the plugin since this took a bit of digging

I ran it and the claims come through as expected and the email is encoded as expected (although the mockserver doesn't handle this)

zloom commented:
Looks good to me, thanks for contribution!
